### PR TITLE
Add resource manager and utils for gatewayroute

### PR DESCRIPTION
**Pull request by fawadkhaliq**
*Description of changes:*
Add resource manager and util helpers for Gateway Route

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/gatewayroute/resource_manager.go [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+type defaultResourceManager struct {
+	k8sClient          client.Client
+	appMeshSDK         services.AppMesh
+	referencesResolver references.Resolver
+	accountID          string
+	log                logr.Logger
+}
+
+func (m *defaultResourceManager) Reconcile(ctx context.Context, gr *appmesh.GatewayRoute) error {
+	ms, err := m.findMeshDependency(ctx, gr)
```
Comment (achevuru) [inline pkg/gatewayroute/resource_manager.go:62]: Is this necessary? wouldn't the below find + validate virtualGatewayDependency validate the Mesh as well? VirtualGateway wouldn't be up and active if the corresponding mesh is missing right?

File: pkg/gatewayroute/resource_manager.go [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+	}
+
+	return m.updateCRDGatewayRoute(ctx, gr, sdkGR)
+}
+
+func (m *defaultResourceManager) Cleanup(ctx context.Context, gr *appmesh.GatewayRoute) error {
+	ms, err := m.findMeshDependency(ctx, gr)
+	if err != nil {
+		return err
+	}
```
Comment (achevuru) [inline pkg/gatewayroute/resource_manager.go:107]: So, if there is no virtualgateway associated with these set of gateway routes then can't we go ahead and delete it? Is there a reason we block the delete for a mesh being active?

File: pkg/gatewayroute/resource_manager.go [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+
+func (m *defaultResourceManager) updateCRDGatewayRoute(ctx context.Context, gr *appmesh.GatewayRoute, sdkGR *appmeshsdk.GatewayRouteData) error {
+	oldGR := gr.DeepCopy()
+	needsUpdate := false
+	if aws.StringValue(gr.Status.GatewayRouteARN) != aws.StringValue(sdkGR.Metadata.Arn) {
+		gr.Status.GatewayRouteARN = sdkGR.Metadata.Arn
+		needsUpdate = true
+	}
+
+	if aws.Int64Value(gr.Status.ObservedGeneration) != gr.Generation {
```
Comment (achevuru) [inline pkg/gatewayroute/resource_manager.go:298]: Should it be only when ObservedGeneration < Generation?

Comment (fawadkhaliq) [reply pkg/gatewayroute/resource_manager.go:298]: Technically yes as this covers that too. Shouldn't cause any issues either way. Do you see any issues with the current approach?

File: pkg/gatewayroute/resource_manager.go (hunk 2 above)
Comment (achevuru) [reply pkg/gatewayroute/resource_manager.go:107]: No, my question is why do we need to check if there is an active mesh in the namespace before deleting gateway routes? Shouldn't we just check for virtual gateway? I can just choose to remove VG and GR and should be able to continue to use the mesh like how we do currently, right?

File: pkg/gatewayroute/resource_manager.go (hunk 3 above)
Comment (achevuru) [reply pkg/gatewayroute/resource_manager.go:298]: I mean, once observedGeneration becomes > generation not sure how they can ever catch up. We will end up updating the resource every time even when there is no need.

**Pull request merged**